main/src/ca/uwaterloo/flix/language/phase/Deriver.scala

```diff
@@ -336,6 +338,214 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
       KindedAst.MatchRule(pat, guard, exp)
   }
 
+  /**
+    * Creates an Order instance for the given enum.
+    *
+    * {{{
+    * enum E[a] with Order {
+    *   case C1
+    *   case C2(a)
+    *   case C3(a, a)
+    * }
+    * }}}
+    *
+    * yields
+    *
+    * {{{
+    * instance Order[E[a]] with Order[a] {
+    *   pub def compare(x: E[a], y: E[a]): Comparison = {
+    *     let indexOf = e -> match e {
+    *       case C0(_) -> 0
+    *       case C1(_) -> 1
+    *       case C2(_) -> 2
+    *     };
+    *     match (x, y) {
+    *       case (C0, C0) => Comparison.EqualTo
+    *       case (C1(x0), C1(y0)) => Order.compare(x0, y0)
+    *       case (C2(x0, x1), C2(y0, y1)) => Order.compare(x0, y0) `Order.thenCompare` lazy Order.compare(x1, y1)
+    *       case _ => Order.compare(indexOf(x), indexOf(y))
+    *     }
+    *   }
+    * }
+    * }}}
+    */
+  private def mkOrderInstance(enum: KindedAst.Enum, loc: SourceLocation, root: KindedAst.Root)(implicit flix: Flix): KindedAst.Instance = enum match {
+    case KindedAst.Enum(_, _, _, tparams, _, cases, _, sc, _) =>
+      val orderClassSym = PredefinedClasses.lookupClassSym("Order", root)
+      val compareDefSym = Symbol.mkDefnSym("Order.compare")
+
+      val param1 = Symbol.freshVarSym("x", loc)
+      val param2 = Symbol.freshVarSym("y", loc)
+      val exp = mkCompareImpl(enum, param1, param2, loc, root)
+      val spec = mkCompareSpec(enum, param1, param2, loc, root)
+
+      val defn = KindedAst.Def(compareDefSym, spec, exp)
+
+      // Add a type constraint to the instance for any non-wild param with kind Star
+      val tconstrs = tparams.collect {
```

Comment:
I think this should be a helper method. (Also for the other cases where it is used).